package Epsilon5;

enum Weapon {
    Pistol = 0;
    Machinegun = 1;
    Shotgun = 2;
}

message Player {
    required int32 id = 1;

    required int32 x = 2;
    required int32 y = 3;

    optional double vx = 4;
    optional double vy = 5;

    optional double angle = 6;
    optional string name = 7;
    optional uint32 hp = 8;
    required bool team = 9;
    optional uint32 ping = 10;
}

message Bullet {
    required int32 x = 1;
    required int32 y = 2;

    optional double vx = 3;
    optional double vy = 4;
    enum Type {
        ARBUZ = 0;
        LITTLE_BULLET = 1;
        ROCKET = 2;
    }
    optional Type bullet_type = 5 [default = ARBUZ];
    required bool team = 6;
}

// Object item
// [server -> client]
// If no any optional values passed, client should remove this object,
// else client only updates'em
message Object {
    required uint32 id = 1;
    optional int32 resource_id = 2;
    optional int32 x = 3;
    optional int32 y = 4;
    optional double angle = 5;
}

message RespPoint {
    required uint32 id = 1;
    required int32 x = 2;
    required int32 y = 3;
    required int32 team = 4;
    required bool is_main = 5;
}

message PlayerStat {
    required int32 id = 1;
    required int32 score = 2;
    required int32 kills = 3;
    required int32 deaths = 4;
}

message PlayerInfo {
    optional int32 id = 1;
    optional string map = 2;
    optional int32 weapon = 3;
    optional int32 bullets = 4;
    optional int32 cage = 5;
}

message World {
    repeated Player players = 1;
    repeated Bullet bullets = 2;
    repeated Object objects = 3;
    repeated RespPoint resp_points = 4;
    required uint32 packet_number = 5;
    repeated PlayerStat players_stat = 6;
    optional PlayerInfo player_info = 7;
}

message Control {
    message KeyStatus {
        required bool keyUp = 1;
        required bool keyDown = 2;
        required bool keyLeft = 3;
        required bool keyRight = 4;

        required bool keyAttack1 = 5;
        required bool keyAttack2 = 6;
    }

    required KeyStatus keyStatus = 1;
    required double angle = 2;
    required Weapon weapon = 3;
    required uint32 packet_number = 4;
    optional bool need_full = 5;
}

message Auth {
    required string name = 1;
    optional string password = 2;
}
